tensorflow/java/src/test/java/org/tensorflow/op/ScopeTest.java

```diff
@@ -0,0 +1,265 @@
+      }
+
+      // Non-root scopes have a less restrictive constraint.
+      assertEquals("a/_/hello", root.withSubScope("a").withSubScope("_").makeOpName("hello"));
+    }
+  }
+
+  @Test
+  public void basic() {
+    try (Graph g = new Graph()) {
+      Scope s = new Scope(g);
+      Const c1 = Const.create(s, 42);
+      assertEquals("Const", c1.output().op().name());
+      Const c2 = Const.create(s, 7);
+      assertEquals("Const_1", c2.output().op().name());
+      Const c3 = Const.create(s.withName("four"), 4);
+      assertEquals("four", c3.output().op().name());
+      Const c4 = Const.create(s.withName("four"), 4);
+      assertEquals("four_1", c4.output().op().name());
+    }
+  }
+
+  @Test
+  public void hierarchy() {
+    try (Graph g = new Graph()) {
+      Scope root = new Scope(g);
+      Scope child = root.withSubScope("child");
+      assertEquals("child/Const", Const.create(child, 42).output().op().name());
+      assertEquals("child/four", Const.create(child.withName("four"), 4).output().op().name());
+    }
+  }
+
+  @Test
+  public void composite() {
+    try (Graph g = new Graph();
+        Session sess = new Session(g)) {
+      Scope s = new Scope(g);
+      Output data = Const.create(s.withName("data"), new int[] {600, 470, 170, 430, 300}).output();
+
+      // Create a composite op with a customized name
+      Variance var1 = Variance.create(s.withName("example"), data);
+      assertEquals("example/variance", var1.output().op().name());
+
+      // Confirm internally added ops have the right names.
+      assertNotNull(g.operation("example/squared_deviation"));
+      assertNotNull(g.operation("example/Mean"));
+      assertNotNull(g.operation("example/zero"));
+
+      // Same composite op with a default name
+      Variance var2 = Variance.create(s, data);
+      assertEquals("variance/variance", var2.output().op().name());
+
+      // Confirm internally added ops have the right names.
+      assertNotNull(g.operation("variance/squared_deviation"));
+      assertNotNull(g.operation("variance/Mean"));
+      assertNotNull(g.operation("variance/zero"));
+
+      // Verify correct results as well.
+      Tensor result = sess.runner().fetch(var1.output()).run().get(0);
+      assertEquals(21704, result.intValue());
+      result = sess.runner().fetch(var2.output()).run().get(0);
+      assertEquals(21704, result.intValue());
+    }
+  }
+
+  // "handwritten" sample operator classes
+  private static final class Const {
+    private final Output output;
+
+    private static Const create(Scope s, Object v) {
+      try (Tensor value = Tensor.create(v)) {
+        return new Const(
+            s.graph()
+                .opBuilder("Const", s.makeOpName("Const"))
+                .setAttr("dtype", value.dataType())
+                .setAttr("value", value)
+                .build()
+                .output(0));
+      }
+    }
+
+    private Const(Output o) {
+      output = o;
+    }
+
+    public Output output() {
```

Comment:
Should we remove these `private` and `public` method qualifiers from the sample ops? (They are only needed for the tests and since they are inner classes, the test code is able to call them anyway (e.g., `Const.create` is private, but still callable - so the distinction between `private` and `public` here is immaterial anyway?)

Alternatively, make the `create` factory function `public` to be closer to where we want to head anyway?